### ISSUE TITLE
How to draw layers with ordering priority?

### ISSUE THREAD

**Issue by leo-costa** [Mapsui/Mapsui#598]
Hi 

I know Mapsui draw layers based on the order in which they were added to the map layer collection.
I have a user case now where some layers must be dynamically inserted into the map. Let's say there are 5 layers that needs to be inserted at a specific location. This is my current logic 

1. Have a 'dummy / marker' layer with a name inserted ahead of time.
2. When a dynamic layer comes, I need to find where this dummy layer is using ```FindLayer``` method
3. I also need to find the index of this dummy layer
4. Using the information above I insert my dynamic layer after the dummy one using ```Insert``` method from ```Map.Layers```

Is there a better way of doing this ? Perhaps a layer ordering could be implemented instead of just iterating though layers when drawing ?

**Comment by pauldendulk** [comment]
There is no real easy way to do that right now. We could try to think of something better.

**Comment by pauldendulk** [comment]
@leo-costa This should be improved. I am trying to get a better picture of the user's perspective. The LayerCollection also has Insert and Move methods but these are now sufficient. What kind of method would help you? 

For instance:
```InsertBefore(ILayer referenceLayer, ILayer layerToInsert)```
```InsertAfter(ILayer referenceLayer, ILayer layerToInsert)```

**Comment by pauldendulk** [comment]
We also need to find a proper solution to reference a layer. We could reference to an instance, like above, it would do reference compare. But it the layers are read from disk this would not help. So we need some other identifier. 
- BaseLayer has an Id which is unique as long as layers are derived from BaseLayer, but a user could implement it's own base layer. 
- We could enforce 'Name' to be unique. But if data is read from disk or imported this might be problematic. It would not be possible to import certain data. Perhaps we could also add a Title which is not required to be unique.

**Comment by pauldendulk** [comment]
We could also add an Order field managed by the user, but this is not always easy. If you want to insert something you need to update the Order fields of the layers following the inserted layer.

We could also manage the order field from the LayerCollection. Basically that is the layer index. Not sure if this would help. If we can add some extra methods so that the user does not need to bother with the Order or Index that would be best.

**Comment by pauldendulk** [comment]
It would also be interesting to know how other libraries solve this.

**Comment by charlenni** [comment]
I looked into this. 

Google Maps API for JavaScript does it [this](https://developers.google.com/maps/documentation/javascript/customoverlays) way. There are more than one layer type. 

[Mapbox](https://docs.mapbox.com/mapbox-gl-js/example/geojson-layer-in-stack/) could place a layer relative to another.

For me it is now ok, when I listen to the changed event and than remove my layers and readd them. Perhaps we could add some functions like `AddLayerAbove(ILayer)` and `AddLayerBelow(ILayer)`. With this it is clear, where this layer should be placed.

**Comment by charlenni** [comment]
@pauldendulk I see, that you suggested this before. I think, your suggested names are even better. Perhaps a `InsertOnTop(ILayer layerToInsert)` and `InsertOnBottom(ILayer layerToInsert)` would be helpful.

**Comment by leo-costa** [comment]
@pauldendulk [WhirlyGlobe](http://mousebird.github.io/WhirlyGlobe/) might be an example to look how ordering is done.

```InsertBefore(ILayer referenceLayer, ILayer layerToInsert)```
```InsertAfter(ILayer referenceLayer, ILayer layerToInsert)```

Is a good start, but how can ordering be achieved without holding a reference to a layer ?
One idea could be to order layers based on a property before iterating though the layers.

For instance, If there was a property ```int Priority {get; set;} = 0``` and the layers never modified this property, then they would be draw in the order they were inserted. 

**Example:**
If I had the layers added to the map collection in the following order.

1) Layer 1 : Priority = 0
2) Layer 2 : Priority = 2
3) Layer 3 : Priority = 2
4) Layer 4 : Priority = 1

The drawing priority would be:

1) Layer 3   ----------> Top Layer
2) Layer 2
3) Layer 4
4) Layer 1    ----------> Bottom Layer

**Comment by pauldendulk** [comment]
@leo-costa So what would happen if you want to insert a layer at position 3? Will there be two layers with Priority 3? Or will the library update the other layers Priority? Or will the user update the other layer's Priority, if so, the user will still need a reference.

**Comment by pauldendulk** [comment]
@leo-costa Does mapply have a documentation section on ordering?

**Comment by charlenni** [comment]
@pauldendulk As you could see in the example, there could be more than one layer of the same priority. The last added layer with a given priority is the top most layer of this group. So, if you add a layer with priority 3, it would be drawn on top of all layers with lower priorities (like 0, 1, 2) and on top of all layers with same priority. With this, you could build groups of layers, which should stay on top of other layers, but the ordering inside of the group isn't important, or you could make a more aligned ordering, where you give each layer a certain priority.

We could call it Priority or Level.

**Comment by pauldendulk** [comment]
I see. I had not read the sample properly. Perhaps this is easy for the user. I am bit worried that the implicit order within a Priority group could become important and would be confusing.

**Comment by charlenni** [comment]
Yes, but that is the same as now, because the ordering is already important. The Priority is only a auxiliary construction. If you want, you could use the normal system with `InsertBefore()` and `InsertAfter()`.

**Comment by leo-costa** [comment]
> I am bit worried that the implicit order within a Priority group could become important and would be confusing.

@pauldendulk I think that by definition if an item within a priority group needs to be displayed in a certain order, then that item needs to have higher priority then the rest of the group.

> Does mapply have a documentation section on ordering?

@pauldendulk  http://mousebird.github.io/WhirlyGlobe/reference/ios_2_5/Classes/MaplyViewControllerLayer.html#/c:objc(cs)MaplyViewControllerLayer(py)drawPriority

Perhaps using a priority queue for the underlying map layer collection if better suited ?

**Comment by pauldendulk** [comment]
> I think that by definition if an item within a priority group needs to be displayed in a certain order, then that item needs to have higher priority then the rest of the group.

Agreed. But if a certain user really needs a specific order over all layers the Priority solution would be an unhandy one. Where you have to update the Priority of all the layers that have to be drawn after the one you are inserting. InsertAfter and InsertBefore would be easier.

What do you mean with the part about 'priority queue'?

**Comment by leo-costa** [comment]
@pauldendulk  The current drawing priority, outside of adding layers in the correct order, is on ```VisibleFeatureIterator.IterateLayers```.  The ```layers``` parameters is currently taking in ```LayerCollection``` which is backed by a ```List<ILayer>```. What I'm proposing is to change  the backing collection of ```LayerCollection```  from a ```List<ILayers>``` to a priority collection of ILayers.

This allows current code using mapsui to work without breaking changes since all layers would have the default ```int Priority {get;set;} = 0```  and insertion ordering would be kept.

Alternatively, on ```VisibleFeatureIterator.IterateLayers```, on the forearch we can do ```layers.OrderByDescending(x=> x.Priority)```, but I think this is a less elegant solution.

**Comment by pauldendulk** [comment]
> from a List<ILayers>to a *priority collection of ILayers.*

I still don't understand what you mean here. So, this is not a List<ILayers>? So what is it? Do you mean we have to sort the layers on insert instead of while iterating in the VisibleFeatureIterator.

**Comment by pauldendulk** [comment]
Another perspective. Some map components have layer 'groups'. This is basically what you want. Where the group is all that matters and the specific ordering within a group is ignored. Maybe we can get some inspiration from that perspective.

**Comment by charlenni** [comment]
@pauldendulk Do you looked into #626. I called it Level instead of Priority, because I think it's the better word for this.

**Comment by leo-costa** [comment]
>I still don't understand what you mean here. So, this is not a List? 

I might not have been very clear. In the end it might as well be a List but ordering needs to happen when inserting a layer.

>Do you mean we have to sort the layers on insert instead of while iterating in the VisibleFeatureIterator.

Yes.

**Comment by pauldendulk** [comment]
@charlenni I would prefer Order. Level is in Mapsui often used for the Level of a tile piramide and I associate Priority with time, like the first to arrive, while this is a special thing. Order is often used as a field in database tables.

Also I am not ready to choose yet. I would like to have the best of both worlds. Inserting an item relative to another layer could be handy too.

**Comment by pauldendulk** [comment]
@charlenni In your PR the Level is set when creating a layer. Perhaps users also want to change the Level later on.

**Comment by charlenni** [comment]
Order is also ok. I have no problem with it. And yes, Priority is something associated with time.

Yes, changing the order later could happen, but what to do than? Change the position in the layers list too? And when yes, where should I get this list from?

**Comment by charlenni** [comment]
If you see in my PR, you could add a layer also by `InsertBefore()` and `InsertAfter()`. So you could use the LayersCollection without any order and make this order by hand, or you could use a order index and sort the LayersCollection by this. And yes, you could both at the same time.

**Comment by pauldendulk** [comment]
You could use both but it feels a bit awkward to combine them. You could insert a layer with Order 0 after a layer with Order 1 and then the inserted layer would still be drawn below the layer it is inserted after.

We would use different systems for individual ordering and grouped ordering. That doesn't feel right.

Another question I have is if we should manage ordering in layer at all. Maybe it should all be the responsibility of the LayerCollection.

**Comment by charlenni** [comment]
The problem in workflow is, that you e.g. want to add a layer over all this polygon pin layers, but below the pin layers. The question is, how to do this. If you use InsertBefore/InsertAfter, you have to know, which is the right layer. If you have a better idea for such a scenario...

**Comment by pauldendulk** [comment]
I was now thinking of this.

The ILayer itself has no ordering information. Only the LayerCollection has. Within the LayerCollection there is actual grouping for what we referred to as Order (which I now think might be called Group to introduce a fourth term  :) ). Internally it is not a List<ILayer> but a IDictionary<int, List<ILayer>>. So a list of lists. Since there is no order information on ILayer everything needs to be arranged through methods of LayerCollection. Some of these methods will have an optional Group parameter. 

Relative methods like ```InsertAfter``` will not have the Group parameter. They will be added to the same group as the reference layer. ```Add``` will have an option Group. When left out it will be Group 0. 
```InsertOnTop``` will have an option Group. (perhaps there is no use for ```Add``` if we have ```InsertOnTop```). The same for ```InsertOnBottom```. Perhaps adding to bottom and top is not inserting, so they should be called ```AddToTop``` and ```AddTopBottom```.

**Comment by pauldendulk** [comment]
This is the interface for LayerCollection I now have in mind

```c#
    public interface ILayerCollection
    {
        void Clear(); // removes all layers
        void Clear(int group); // removes all layers in a group
        void AddToTop(ILayer layer, int group = 0);
        void AddToBottom(ILayer layer, int group = 0);
        void InsertBefore(ILayer referenceLayer, ILayer layerToInsert);
        void InsertAfter(ILayer referenceLayer, ILayer layerToInsert);
        void MoveUp(ILayer layer); // if it is the top layer in the group it will become the bottom layer in the group above
        void MoveDown(ILayer layer); // if it is the bottom layer in the group it will become the top layer in the group above
        ILayer Remove(ILayer layer); // Returns the removed layers so that it can be used for moving
    }
```

**Comment by charlenni** [comment]
Does ILayer has a Group property?

MoveUp and MoveDown should have a flag, if a Group change is allowed.

Perhaps a MoveTop and MoveBottom, which moves the layer to the top or bottom of the group?

**Comment by pauldendulk** [comment]
> Does ILayer has a Group property?

No. Do you see a disadvantage? 

> MoveUp and MoveDown should have a flag, if a Group change is allowed.

Makes sense.

> Perhaps a MoveTop and MoveBottom, which moves the layer to the top or bottom of the group?

Ah, yes, Perhaps with the same flag as MoveUp/MoveDown to limit it to the group.

**Comment by leo-costa** [comment]
> No. Do you see a disadvantage?

Perhaps you might want to change ordering or a layer by just changing the property. How would an order / group or a layer change once it was already *inserted* into the collection ?

**Comment by pauldendulk** [comment]
> How would an order / group or a layer change once it was already inserted into the collection?

Good question. My initial idea was to use Remove, which returns the layer removed. So, 

```layers.AddToTop(layers.Remove(myLayer, newOrder):```

But an alternative would be to add add MoveToTop which takes an Order parameter like the AddToTop.

**Comment by pauldendulk** [comment]
I now have this.

```c#
    public interface ILayerCollection
    {
        void Clear(); // removes all layers (or should the default be group 0)
        void Clear(int group); // removes all layers in a group
        void AddToTop(ILayer layer, int group = 0);
        void AddToBottom(ILayer layer, int group = 0);
        void MoveToTop(ILayer layerToMove, int group = 0);
        void MoveToBottom(ILayer layerToMove, int group = 0);		
        void MoveUp(ILayer layerToMove, keepWithinGroup = false); // if it is the top layer in the group it will become the bottom layer in the group above, but not if keepWithinGroup is true
        void MoveDown(ILayer layerToMove, keepWithinGroup = false); // if it is the bottom layer in the group it will become the top layer in the group above, but not if keepWithinGroup is true
        void InsertBefore(ILayer referenceLayer, ILayer layerToInsert);
        void InsertAfter(ILayer referenceLayer, ILayer layerToInsert);
        ILayer Remove(ILayer layer); // Returns the removed layers so that it can be used for moving
    }
```

Perhaps this should be in 3.0.
